torch/cuda/amp/amp_scaler.py

```diff
@@ -0,0 +1,394 @@
+                return val * stash[0].get(val.device)
+            elif isinstance(val, container_abcs.Iterable):
+                return type(val)(apply_scale(v) for v in val)
+            else:
+                raise ValueError("outputs must be a Tensor or an iterable of Tensors")
+
+        return apply_scale(outputs)
+
+    def _unscale_grads(self, optimizer, inv_scale, found_inf, allow_fp16=False):
+        per_device_inv_scale = _MultiDeviceReplicator(inv_scale)
+        per_device_found_inf = _MultiDeviceReplicator(found_inf)
+
+        for group in optimizer.param_groups:
+            for param in group["params"]:
+                if param.grad is not None:
+                    if (not allow_fp16) and param.grad.dtype == torch.float16:
+                        raise ValueError("Attempting to unscale FP16 gradients.  If you want to check for "
+                                         "infs/nans without unscaling, use optimizer.check_infs() instead.")
+                    else:
+                        torch._amp_non_finite_check_and_unscale_(param.grad,
+                                                                 per_device_found_inf.get(param.grad.device),
+                                                                 per_device_inv_scale.get(param.grad.device))
+
+        return per_device_found_inf._per_device_tensors
+
+    def unscale(self, optimizer):
+        """
+        Divides ('unscales') the optimizer's gradient tensors by the scale factor.
+
+        If :meth:`unscale` is not called explicitly then gradients will be
+        automatically unscaled during :meth:`step`.
+
+        :meth:`unscale` can be called explicitly to manipulate a network's
+        gradients after backward and before :meth:`step`.
+
+        Simple example, using :meth:`unscale` to enable clipping of unscaled gradients::
+
+            ...
+            scaler.scale(loss).backward()
+            scaler.unscale(optimizer)
+            torch.nn.utils.clip_grad_norm_(model.parameters(), max_norm)
+            scaler.step(optimizer)
+            scaler.update()
+
+        If this instance of :class:`AmpScaler` is not enabled, :meth:`unscale` is a no-op.
+
+        Arguments:
+            optimizer (torch.optim.Optimizer):  Optimizer that owns the gradients to be unscaled.
+
+        .. note::
+            :meth:`unscale` does not incur a CPU-GPU sync.
+
+        .. warning::
+            :meth:`unscale` should only be called once per optimizer per step,
+            and only after all gradients for that optimizer's owned parameters
+            have been accumulated.
+        """
+        if not self._enabled:
+            return
+
+        assert self._scale is not None, self._scale_not_initialized_error("unscale")
+
+        optimizer_state = self._per_optimizer_states[id(optimizer)]
+
+        if "unscaled" in optimizer_state:
+            raise RuntimeError("unscale() has already been called on this optimizer this iteration.")
```

Comment:
(this comment doesn't really go here, but...)

Greg: if I had the data model explained to me upfront it would have really helped me understand what's going on in this PR.